File: lesson-08/java/MyLinkerListRun.java
public class MyLinkerListRun {
    public static void main(String[] args) {
        MyLinkerList ll = new MyLinkerList();
        for (int i = 1; i <= 5; i++) {
            ll.append(i * 10);
        }
        System.out.println(ll);
        ll.remove(20);
        System.out.println(ll);
    }
}

class MyLinkerList {
    private Node head;
    private int size;

    private static class Node {
        int value;
        Node next;

        Node(int data) {
            this.value = data;
            this.next = null;
        }
    }

    public MyLinkerList() {
        head = null;
        size = 0;
    }

    public void append(int item) {
        Node node = new Node(item);
        if (head == null) {
            head = node;
        } else {
            Node current = head;
            while (current.next != null) {
                current = current.next;
            }
            current.next = node;
        }
        size++;
    }

    public void appendRange(int... items) {
        for (int value : items) {
            append(value);
        }
    }

    public void remove(int data) {
        if (head == null)
            return;

        if (head.value == data) {
            head = head.next;
            size--;
            return;
        }

        Node current = head;
        while (current.next != null && current.next.value != data) {
            current = current.next;
        }

        if (current.next != null) {
            current.next = current.next.next;
            size--;
        }
    }

    public void removeAt(int position) {
        if (position < 0 || position >= size)
            return;

        if (position == 0) {
            head = head.next;
        } else {
            Node current = head;
            for (int i = 0; i < position - 1; i++) {
                current = current.next;
            }
            current.next = current.next.next;
        }
        size--;
    }

    public int get(int index) {
        if (index < 0 || index >= size)
            throw new IndexOutOfBoundsException();

        Node current = head;
        for (int i = 0; i < index; i++) {
            current = current.next;
        }
        return current.value;
    }

    public int size() {
        return size;
    }

    public boolean contains(int data) {
        Node current = head;
        while (current != null) {
            if (current.value == data) {
                return true;
            }
            current = current.next;
        }
        return false;
    }

    @Override
    public String toString() {
        Node current = head;
        StringBuilder out = new StringBuilder();
        while (current != null) {
            out.append(current.value + " ");
            current = current.next;
        }
        return "[" + out.toString().trim() + "]";
    }
}

File: lesson-08/js/MyLinkerListRun.js
class Node {
    constructor(data) {
        this.value = data;
        this.next = null;
    }
}

class MyLinkerList {
    constructor() {
        this.head = null;
        this.size = 0;
    }

    append(item) {
        const node = new Node(item);
        if (this.head === null) {
            this.head = node;
        } else {
            let current = this.head;
            while (current.next !== null) {
                current = current.next;
            }
            current.next = node;
        }
        this.size++;
    }

    appendRange(...items) {
        for (const value of items) {
            this.append(value);
        }
    }

    remove(data) {
        if (this.head === null) return;

        if (this.head.value === data) {
            this.head = this.head.next;
            this.size--;
            return;
        }

        let current = this.head;
        while (current.next !== null && current.next.value !== data) {
            current = current.next;
        }

        if (current.next !== null) {
            current.next = current.next.next;
            this.size--;
        }
    }

    removeAt(position) {
        if (position < 0 || position >= this.size) return;

        if (position === 0) {
            this.head = this.head.next;
        } else {
            let current = this.head;
            for (let i = 0; i < position - 1; i++) {
                current = current.next;
            }
            current.next = current.next.next;
        }
        this.size--;
    }

    get(index) {
        if (index < 0 || index >= this.size) throw new Error("Index out of bounds");

        let current = this.head;
        for (let i = 0; i < index; i++) {
            current = current.next;
        }
        return current.value;
    }

    size() {
        return this.size;
    }

    contains(data) {
        let current = this.head;
        while (current !== null) {
            if (current.value === data) {
                return true;
            }
            current = current.next;
        }
        return false;
    }

    toString() {
        let current = this.head;
        const out = [];
        while (current !== null) {
            out.push(current.value);
            current = current.next;
        }
        return `[${out.join(" ")}]`;
    }
}

const main = () => {
    const ll = new MyLinkerList();
    for (let i = 1; i <= 5; i++) {
        ll.append(i * 10);
    }
    console.log(ll.toString());
    ll.remove(20);
    console.log(ll.toString());
}

main();
